docs_src/design/adr/security/0009-Secure-Bootstrapping.md

```diff
@@ -0,0 +1,453 @@
+# Secure Bootstrapping of EdgeX
+
+## Status
+
+** proposed **
+
+## Context
+
+This ADR seeks to define a mechanism for secure bootstrapping of EdgeX.
+This ADR is primarily concerned with solving the problem in OCI container environments,
+but the solutions proposed herein should also be applicable to less-challenging deployment scenarios.
+
+In order to achive the goal of microservice communication security,
+it is necessary to ensure that all EdgeX services are brought up in a secure configuration.
+In Fuji and Geneva releases of EdgeX,
+the security subsystem is bootstrapped through the use of Consul health checks
+to gate the initialization of security services.
+This solution worked for the most part,
+but suffers a critical flaw:
+Consul is used in the EdgeX framework for service location,
+and Consul itself is not configured in a secure mode.
+We need a bootstrapping mechanism that ensures that all services,
+including Conul, are brought up in a secure configuration.
+It is impossible,
+once certain services have been started (such as Consul),
+to go back and secure them on-the-fly.
+
+Secure bootstrapping is needed for any installation of EdgeX,
+but the problem is more profound for a container-based deployment of EdgeX
+because a docker-compose script does not have an explicit installation phase.
+The information in this ADR is generally applicable to the install phase of an EdgeX snap,
+or any manual installation of EdgeX.
+
+A security-disabled installation of EdgeX is not required to follow this guidance.
+
+### Items requiring secure bootstrapping
+
+#### Secrets seed
+
+The secrets seed is a hardware-protected secret value that is used for key derivation.
+The **seed** can be used as input key material to a key deriviation function
+that generates an unlimited number of derived keys from an initial secret,
+all of which appear to be random.
+
+Derived keys can be used to protect bootstrapping secrets
+that cannot otherwise be protected by the EdgeX secret store.
+
+#### Consul
```

Comment:
I feel like this section has strayed from the consul documentation nomenclature or my comprehension has.

> encryption key used by Serf
This is referencing gossip encryption correct ( https://www.consul.io/docs/agent/encryption#gossip-encryption )?

> For TLS: a CA certificate, TLS certificate (chain), and TLS private key
This is referencing RPC encryption with TLS? ( https://www.consul.io/docs/agent/encryption#rpc-encryption-with-tls )

currently neither of these settings are enabled with edgex currently right? My nightly build shows:
> Encrypt: Gossip: false, TLS-Outgoing: false, TLS-Incoming: false, Auto-Encrypt-TLS: false

Reply:
Correct.  Neither of these features are currently enabled.  Thus, it is possible for a rogue network agent to join itself to consul and start reading or writing the key-value store.  Also consul is subject to man-in-the-middle attacks that can replace critical configuration information (such as service location).